kernel-rs/src/virtio/virtio_disk.rs

```diff
@@ -352,6 +375,210 @@ impl VirtioDisk {
         guard.wakeup(ctx.kernel());
     }
 
+    // This method writes the data in the given buffers to disk sequentially, therefore,
+    // minimizing seek-time overhead of the disk. It is designed for using with `Lfs`.
+    // Calling this method when using `Ufs` is effectively not different from calling
+    // `VirtioDisk::rw` and may incur additional performance overhead.
+    fn write_seq(
+        guard: &mut SleepableLockGuard<'_, Self>,
+        barray: &mut [Option<Buf>; MAX_SEQ_WRITE],
+        ctx: &KernelCtx<'_, '_>,
+    ) {
+        // Maximum # of blocks to be sequentially written to the disk must be less than
+        // the total # of available descriptors excluding two descriptors, which will be
+        // allocated for the header and tailer descriptors.
+        const_assert!(MAX_SEQ_WRITE + 2 <= NUM);
+
+        let mut darray: [Option<Descriptor>; MAX_SEQ_WRITE] = array![_ => None; MAX_SEQ_WRITE];
+        let mut prev_blk_idx = 0usize;
+        let mut next_write = 0usize;
+        let mut header: Option<Descriptor> = None; // To reserve a header descriptor.
+        let mut tailer: Option<Descriptor> = None; // To reserve a tailer descriptor.
+
+        for i in 0..MAX_SEQ_WRITE {
+            if barray[i].is_some() {
+                let desc = loop {
+                    // Allocate a header and tailer descriptor if not already allocated.
+                    // Invariant: `header` and `tailer` are always allocated together.
+                    if header.is_none() {
+                        // `None` means this is the beginning of a new descriptor chain.
+                        match guard.get_pin_mut().alloc() {
+                            None => {
+                                guard.sleep(ctx);
+                                continue;
+                            }
+                            idx => header = idx,
+                        }
+                        match guard.get_pin_mut().alloc() {
+                            None => {
+                                guard.get_pin_mut().as_mut().free(header.unwrap());
+                                header = None;
+                                continue;
+                            }
+                            idx => tailer = idx,
+                        }
+                    }
+
+                    // Allocate a descriptor for the buffer.
+                    match guard.get_pin_mut().alloc() {
+                        Some(idx) => break idx,
+                        // Back down to prevent deadlock
+                        None => {
+                            if Self::finalize_write_seq(
+                                guard,
+                                &mut darray[next_write..=prev_blk_idx],
+                                &mut barray[next_write..=prev_blk_idx],
+                                &mut header,
+                                &mut tailer,
```

Comment:
- I am sorry. I do not understand why we should not use the buffers this way.
- `header` and `tailer` should be allocated before buffer descriptors because they are necessary for forming a disk request. Otherwise, I think it will result in deadlock.

Reply:
I meant that `tailer` should be written at the **last allocated descriptor** if we are planning to use the `VIRTIO_F_IN_ORDER` feature.
We don't need to care about this since we're not using `VIRTIO_F_IN_ORDER` anyway, but if we want to, we may need to first allocate 2 descriptors -> only write the `header` at first one -> allocate a single descriptor for a `Buf` (repeatedly) -> finally write the `tailer` (at the last allocated descriptor).

Reply:
Understood. Thank you very much.